protocol/auditlog.go

```diff
@@ -0,0 +1,180 @@
+// for "catching up" on a history in case an auditor misses epochs
+func (l *ConiksAuditLog) Insert(addr string, signKey sign.PublicKey,
+	oldSTRs map[uint64]*DirSTR, latestSTR *DirSTR) error {
+
+	// error if we want to create a new entry for an addr we already know
+	if l.IsKnownDirectory(addr) {
+		return ErrAuditLog
+	}
+
+	// create the new directory history
+	h := newDirectoryHistory(signKey, latestSTR)
+
+	startEp := uint64(0)
+	endEp := latestSTR.Epoch
+
+	// add each old STR into the history
+	for ep := startEp; ep < endEp; ep++ {
+		str := oldSTRs[ep]
+		if str == nil {
+			return ErrMalformedDirectoryMessage
+		}
+		h.snapshots[ep] = str
+	}
+
+	l.histories[addr] = h
+
+	// FIXME: verify the consistency of each new STR
+	return nil
+}
+
+// Update verifies the consistency of a newly observed STR newSTR for
+// the directory addr, and inserts the newSTR into addr's directory history
+// if the checks (i.e. STR signature and hash chain verifications) pass.
+// Update() returns nil if the checks pass, and the appropriate consistency
+// check error otherwise. Update() assumes that Insert() has been called for
+// addr prior to its first call and thereby expects that an entry for addr
+// exists in the audit log l.
+// FIXME: pass Response message as param
+func (l *ConiksAuditLog) Update(addr string, newSTR *DirSTR) error {
+
+	// error if we want to update the entry for an addr we don't know
+	if !l.IsKnownDirectory(addr) {
+		return ErrAuditLog
+	}
+
+	h := l.histories[addr]
+
+	if err := verifySTRConsistency(h.signKey, h.latestSTR, newSTR); err != nil {
+		return err
+	}
+
+	// update the latest STR
+	h.snapshots[h.latestSTR.Epoch] = h.latestSTR
+	h.latestSTR = newSTR
+	return nil
+}
+
+// GetObservedSTRs gets the observed STR for the CONIKS directory
+// address for a directory history entry indicated in the
+// AuditingRequest req received from a CONIKS client,
+// and returns a tuple of the form (response, error).
+// The response (which also includes the error code) is supposed to
+// be sent back to the client. The returned error is used by the auditor
+// for logging purposes.
+//
+// A request without a directory address or with an epoch greater than
+// the latest observed epoch of this directory is considered malformed,
+// and causes GetObservedSTRs() to return a
+// message.NewErrorResponse(ErrMalformedClientMessage) tuple.
+// GetObservedSTRs() returns a message.NewSTRHistoryRange(strs) tuple.
+// strs is a list of STRs for the epoch range [ep,
+// l.histories[req.DirectoryAddr].latestSTR.Epoch], where ep is the epoch for
+// which the client has requested the observed STR; i.e. if ep == the latest epoch,
+// the list returned is of length 1.
+// If the auditor doesn't have any history entries for the requested CONIKS
+// directory, GetObservedSTRs() returns a
+// message.NewErrorResponse(ReqUnknownDirectory) tuple.
+func (l *ConiksAuditLog) GetObservedSTRs(req *AuditingRequest) (*Response,
```

Comment:
> In an earlier iteration we had two separate requests, [...]

I remember this, too. But the main reason I suggested this because we have #173 now. 

> Is the concern here that the client could receive an STR from the auditor than the STR it received from the server during the last monitoring?

Yes.

> The only way this could happen is if the client finishes with monitoring very close to/at the end of an epoch. [...]

I think it could also happen if the `EpochDeadline` is very small, say, 5s, and the client has a high network latency, even if it does monitoring at the beginning of en epoch. But I agree that we shouldn't make things more complicated. What you said makes sense to me.

> Since the client only stores the latest STR from the server, it would do a hash chain verification on the STR range from the auditor and just compare the latest STRs with each other.

Do you mean the client fetches a range of STR from the server, verifies them, and compare the newest STR in the range with the auditors?